src/libraries/System.Private.CoreLib/src/System/Environment.Win32.cs

```diff
@@ -217,31 +217,40 @@ private static string GetFolderPathCore(SpecialFolder folder, SpecialFolderOptio
             // The only SpecialFolderOption defines we have are equivalent to KnownFolderFlags.
 
             string folderGuid;
-
+            string? fallbackEnv = null;
             switch (folder)
             {
                 // Special-cased values to not use SHGetFolderPath when we have a more direct option available.
                 case SpecialFolder.System:
-                    // This assumes the system directory always exists and thus we don't need to do anything special for any SpecialFolderOption.
-                    return SystemDirectory;
+                    if (!string.IsNullOrEmpty(SystemDirectory))
+                        return SystemDirectory;
+
+                    string? windir = Environment.GetEnvironmentVariable("windir");
+                    if (string.IsNullOrEmpty(windir))
+                        return string.Empty;
+
+                    return Path.Combine(windir, "system32");
                 default:
                     return string.Empty;
 
-                // Map the SpecialFolder to the appropriate Guid
                 case SpecialFolder.ApplicationData:
                     folderGuid = Interop.Shell32.KnownFolders.RoamingAppData;
+                    fallbackEnv = "APPDATA";
                     break;
                 case SpecialFolder.CommonApplicationData:
                     folderGuid = Interop.Shell32.KnownFolders.ProgramData;
+                    fallbackEnv = "ProgramData";
                     break;
                 case SpecialFolder.LocalApplicationData:
                     folderGuid = Interop.Shell32.KnownFolders.LocalAppData;
+                    fallbackEnv = "LOCALAPPDATA";
                     break;
                 case SpecialFolder.Cookies:
                     folderGuid = Interop.Shell32.KnownFolders.Cookies;
                     break;
                 case SpecialFolder.Desktop:
                     folderGuid = Interop.Shell32.KnownFolders.Desktop;
+                    fallbackEnv = "PUBLIC";
```

Comment:
`PUBLIC` env variable does not match desktop path on my machine. Is this the right fallback?